Tenant cluster labelling gsctl (#481)
giantswarm/docs: src/content/reference/gsctl/update-cluster.md

---
title: "gsctl Command Reference: update cluster"
description: The 'gsctl update cluster' command allows the modification of the name of a cluster and its labels.
date: 2020-05-04
type: page
weight: 44
---

# `gsctl update cluster`

The `gsctl update cluster` allows the modification of the name of a cluster and its labels.

> Cluster labelling is only available for clusters with node pools on AWS.

Reviewer: Actually, only on AWS for clusters using releases >v10.x.x.

## Usage

The command is called with the cluster ID or name as a positional argument.
The desired new name can be specified with the `--name` or `-n` flag.
The `--label` flag is used to modify labels.
It can receive either a JSON formatted [JSON Merge Patch, RFC 7386](https://tools.ietf.org/html/rfc7386) or a string specifying label changes.

### Name modification example

```nohighlight
$ gsctl update cluster f01r4 --name "Precious Production Cluster"
```

### Labels modification JSON Merge Patch example

```nohighlight

> $ gsctl update cluster vxvc7 --labels '{"environment": "testing", "locked": null}'

Reviewer: Not digging the syntax (use of JSON) here. Let's come up with something that is more CLI-native.

Author: so only the `key=value` syntax then?

Reviewer: And then I found the syntax described below. When we make that usable, I see no reason to support JSON in the CLI.

Author: What about people using gsctl in automation not wanting to talk directly to the API? I am very much fine with implementing less, but the original story (https://github.com/giantswarm/giantswarm/issues/1693) even mentions usage of tenant cluster labels in CI context

Reviewer: > What about people using gsctl in automation not wanting to talk directly to the API?

They should be able to use `--label` one or several times in the non-JSON way.

Author: We'll do it the multiple `--label` way to get implementation tires kicking..

```

will update the labels of cluster `vxvc7`. It will add (or update depending on prior existence) label `environment=testing` and delete the label with key `locked`.

### Labels modification key=value example

```nohighlight

> $ gsctl update cluster vxvc7 --labels 'environment=testing locked-'

Reviewer: I would strongly suggest use of multiple `--label` flags here (not `--labels`). Also I don't like the custom "minus" syntax, an empty string should be more expressive.

```suggestion
$ gsctl update cluster vxvc7 --label environment=testing --label locked=
```

In spf13/cobra, using a `StringSlice` flag for the above means that the following syntax will also work:

```
$ gsctl update cluster vxvc7 --label environment=testing,locked=
```

Author: This is lifted from `kubectl labels` https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#label

Reviewer 2: I would also like to align with kubectl here as this is also going to be the interaction basis in the future while we open our control plane.

Reviewer: To me, aligning with a very specific kubectl behaviour in a tool that is in many ways different from kubectl is not a goal we should pursue.

That would be different if gsctl was a kubectl plugin already.

```

will update the labels of cluster `vxvc7`. It will add (or update depending on prior existence) label `environment=testing` and delete the label with key `locked`.

## Full argument reference {#arguments}

- `--name` or `-n`: The new cluster name.

> - `--labels`: Specify label updates. Allows to specify label updates either as a [JSON Merge Patch, RFC 7386](https://tools.ietf.org/html/rfc7386) or in form of `'key=value anotherkey=anothervalue'`.

Reviewer: Would need adaptation according to the example above

> Use `key-` to delete the label with key `key`.

Reviewer: ```suggestion
```

> Use `key-` to delete the label with key `key`.

Reviewer: ```suggestion
To remove a label, set it to an empty string.
```

# `gsctl update cluster`

The `gsctl update cluster` allows the modification of the name of a cluster and its labels.

> Cluster labelling is only available for clusters with release version 10.0.0 and above on AWS.

Reviewer: ```suggestion
Cluster labelling is only available for clusters with release version {{% first_aws_nodepools_version %}} and above on AWS.
```
